File: math/happy-number.py
# https://leetcode.com/problems/happy-number/
# tags - number
class Solution:
    def isHappy(self, n: int) -> bool:
        def extractDigits(n):
            digits = []
            while n > 0:
                digits.append(n % 10)
                n = n // 10
            return digits
        
        seen = set()
        while (n != 1 and n not in seen):
            seen.add(n)
            n = sum([d*d for d in extractDigits(n)])
        return n == 1 
        

File: math/happy-number.js
// https://leetcode.com/problems/happy-number/
// tags - math
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
  const set = new Set();
  while (!set.has(n)) {
    set.add(n);
    let digitSqSum = 0;
    while (n > 0) {
      const digit = n % 10;
      digitSqSum += digit * digit;
      n = Math.floor(n / 10);
    }
    n = digitSqSum;
  }
  return n === 1;
};
